### PR TITLE
8298639: Perform I/O operations in bulk for RandomAccessFile

### PR DESCRIPTION
**Pull request by minborg**
This PR suggests improving performance for read and write operations for the longer primitives in `RandomAccessFile`.

These improvements would also propagate to other JDK classes relying on these improved `RandomAccessFile` operations.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298639](https://bugs.openjdk.org/browse/JDK-8298639): Perform I/O operations in bulk for RandomAccessFile

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Contributors
 * Sergey Tsypanov `<stsypanov@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11644/head:pull/11644` \
`$ git checkout pull/11644`

Update a local copy of the PR: \
`$ git checkout pull/11644` \
`$ git pull https://git.openjdk.org/jdk pull/11644/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11644`

View PR using the GUI difftool: \
`$ git pr show -t 11644`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11644.diff">https://git.openjdk.org/jdk/pull/11644.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
@stsypanov is working on the same thing in JDK-8292937 / PR10031.

**Comment by stsypanov** [comment]
As far as I understood https://github.com/openjdk/jdk/pull/10031 was closed because there we doubts about whether it's safe to use shared array

**Comment by AlanBateman** [comment]
> As far as I understand #10031 was closed because there we doubts about whether it's safe to use shared array within multiple methods. Here we use the same approach.

I think that PR was closed as it went inactive but I think where you got to is very similar to what Per's change is now.

RandomAccessFile doesn't specify it but it can't be used from several threads without external synchronization. This is because it maintains a file position that is updated when reading and writing, or it can be changed when seeking to positions in the file. So on the surface, the byte[] should be okay.

**Comment by stsypanov** [comment]
> I think that PR was closed as it went inactive

Yep, I think there I did everything I could, hope this one will be merged as the changes are helpful

**Comment by minborg** [comment]
Before (baseline):

```
Benchmark                             (kiloBytes)  Mode  Cnt      Score     Error  Units
RandomAccessFileBenchmark.readInt               1  avgt   10    425,047 ?   4,391  us/op
RandomAccessFileBenchmark.readInt               5  avgt   10   2116,130 ?  33,072  us/op
RandomAccessFileBenchmark.readLong              1  avgt   10    425,847 ?   3,115  us/op
RandomAccessFileBenchmark.readLong              5  avgt   10   2118,740 ?  10,773  us/op
RandomAccessFileBenchmark.readShort             1  avgt   10    425,054 ?   6,808  us/op
RandomAccessFileBenchmark.readShort             5  avgt   10   2198,914 ? 105,192  us/op
RandomAccessFileBenchmark.writeInt              1  avgt   10   2259,429 ? 111,989  us/op
RandomAccessFileBenchmark.writeInt              5  avgt   10  10927,965 ? 579,343  us/op
RandomAccessFileBenchmark.writeLong             1  avgt   10   2181,974 ? 136,907  us/op
RandomAccessFileBenchmark.writeLong             5  avgt   10  11319,550 ? 408,315  us/op
RandomAccessFileBenchmark.writeShort            1  avgt   10   2283,893 ? 122,679  us/op
RandomAccessFileBenchmark.writeShort            5  avgt   10  11397,006 ? 543,862  us/op
```

After:

```
Benchmark                             (kiloBytes)  Mode  Cnt     Score     Error  Units
RandomAccessFileBenchmark.readInt               1  avgt   10   132,385 ?   2,416  us/op
RandomAccessFileBenchmark.readInt               5  avgt   10   655,828 ?  14,442  us/op
RandomAccessFileBenchmark.readLong              1  avgt   10    66,383 ?   0,763  us/op
RandomAccessFileBenchmark.readLong              5  avgt   10   329,242 ?   9,269  us/op
RandomAccessFileBenchmark.readShort             1  avgt   10   263,399 ?   4,584  us/op
RandomAccessFileBenchmark.readShort             5  avgt   10  1322,607 ?  27,811  us/op
RandomAccessFileBenchmark.writeInt              1  avgt   10   535,133 ?  24,860  us/op
RandomAccessFileBenchmark.writeInt              5  avgt   10  2701,959 ? 115,069  us/op
RandomAccessFileBenchmark.writeLong             1  avgt   10   268,175 ?  13,810  us/op
RandomAccessFileBenchmark.writeLong             5  avgt   10  1313,634 ?  33,461  us/op
RandomAccessFileBenchmark.writeShort            1  avgt   10  1076,690 ?  17,407  us/op
RandomAccessFileBenchmark.writeShort            5  avgt   10  5383,567 ? 250,833  us/op
```

image

File: src/java.base/share/classes/java/io/RandomAccessFile.java [hunk 1]
```diff
@@ -57,6 +57,8 @@
  * than {@code EOFException} is thrown. In particular, an
  * {@code IOException} may be thrown if the stream has been closed.
  *
+ * @implNote This class is not thread safe.
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/RandomAccessFile.java:60]: I don't think this should be an implNote. Maybe for a different issue/PR is that the class description probably needs a paragraph to say that a RAF is not thread safe and requires appropriate synchronization when used by concurrent threads. This leads to discussion about a FileChannel obtained from a RAF as FileChannel defines read/write methods that support cases where threads need to access different parts of a file at the same time. There is a lot more to say on this topic and probably best to keep it separate.

Comment (bplb) [reply src/java.base/share/classes/java/io/RandomAccessFile.java:60]: Nit: Also I think it needs hyphenation: `thread-safe`.

**Review by AlanBateman — APPROVED** [review]

**Review by bplb — APPROVED** [review]

**Pull request closed without merging**